modules/cudaarithm/src/cuda/transpose.cu

```diff
@@ -201,6 +199,9 @@ void cv::cuda::transpose(InputArray _src, OutputArray _dst, Stream& stream)
       else if (!(elemSize%8) && ((elemSize/8)==2))
         nppSafeCall( nppiTranspose_32f_C4R(src.ptr<Npp32f>(), static_cast<int>(src.step),
           dst.ptr<Npp32f>(), static_cast<int>(dst.step), sz) );
+
+      if (!stream)
```

Comment:
Should this be applied to `gridTranspose()` aswell, I can't understand why it was previously missing, surely the result could have been in flight on returns from this function when no stream is passed?

Reply:
I have the opposite question : why is it ever needed ? I let them here, but I don't understand the purpose of this extra synchronization when the default stream (sync by default) is used.

Reply:
Kernel launches are [asynchronous with respect](https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html) with respec to the host.  The [default stream](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html) syncs with other streams by default in legacy mode.

My interpretation is that the OpenCV API works on the assumption that if a stream isn't passed the user wants synchronization.

Reply:
Do you mean that new tha exists both `cudaStreamPerThread` and `cudaStreamLegacy`, people using OpenCV always expect the `cudaStreamLegacy`, thus requiring the `cudaDeviceSynchronize()` ?
If so, I agree.

Reply:
No users of the api will expect functions which they don't pass a stream to, to be synchronous with respect to the host when they return.  This would not be the case with either `cudaStreamPerThread` or `cudaStreamLegacy` as the kernel launch is asynchronous in both cases. 

In the example from the docs

```
1) k_1<<<1, 1, 0, s>>>();
2) k_2<<<1, 1>>>();
3) k_3<<<1, 1, 0, s>>>()
4) ...
```

k_2 waits on k_1 because k_2 is in the legacy default stream, then k_3 waits on the legacy stream.  Because of the specific way this has been set up k_1 and k_2 have finished executing before the call to `k_3<<<1, 1, 0, s>>>()` however the result from k_3 may still be in flight after control has returned to the host when you reach line 4.

Now I haven't used per thread default streams (I always use explicit streams) but my understanding is that if the CUDA_API_PER_THREAD_DEFAULT_STREAM macro was used to enable per thread default streams k_1 would run before k_3 but both would be asynchronous with respect to k_2.  Either way when control returns to the host on line 4 they may still all be in flight.

On the other hand if we have the following which is our case if no stream is passed

```
1) k_1<<<1, 1>>>();
2) k_2<<<1, 1>>>();
3) k_3<<<1, 1>>>();
4) 
```

in either case `cudaStreamPerThread` or `cudaStreamLegacy` all three kernels may still be in flight when control returns to the host on line 4 if we don't explicitly syncronize.

Reply:
What is the problem of the kernel on the default stream being in flight when returning ?

Reply:
The OpenCV cuda API guarantees synchronization when the stream isn't passed to one of its functions.

Reply:
Ok, anyway, I just checked and `gridTranspose()` does ultimately call `cudaDeviceSynchronize()`(see [https://github.com/opencv/opencv_contrib/blob/4.x/modules/cudev/include/opencv2/cudev/grid/detail/transpose.hpp](https://github.com/opencv/opencv_contrib/blob/4.x/modules/cudev/include/opencv2/cudev/grid/detail/transpose.hpp))

Reply:
@cudawarped Do you have open issues with the PR. I want to merge, if you do not mind.

Reply:
@asmorkalov I am not 100% convinced that `isNppiNativelySupported` and `isElemSizeSupportedByNppi` flags are needed.  I can understand that @chacha21 wants a fall back option in case something goes wrong and because of that wants to seperate out the two logic paths but I am not sure if the redundant calls under `isElemSizeSupportedByNppi` make the function harder to maintain, what do you think?

Reply:
Whatever the decision, I will certainly agree : I explained my initial code structure, but I am OK to adapt for a more "OpenCV style"

Reply:
1. Conditions like  `if (!(elemSize%2) && ((elemSize/2)==2))` looks very cryptic. I understand the logic behind it, but  `if (elemSize==4)` definitely more readable.
2. The only reason for fallback I see is 2 channel matrix. All other cases are handled with regular types. I propose to merge "native" support and just "support" in single case and use relevant Npp call. No duplicated/dead branches, mor obvious testing.